src/Core/Services/Cache/DabCacheService.cs

```diff
@@ -0,0 +1,137 @@
+using Microsoft.Extensions.Logging;
+using ZiggyCreatures.Caching.Fusion;
+
+namespace Azure.DataApiBuilder.Core.Services.Cache;
+
+/// <summary>
+/// Service which wraps an internal cache implementation that enables
+/// provided an in-memory cache for the DAB engine.
+/// </summary>
+public class DabCacheService
+{
+    // Dependencies
+    private readonly IFusionCache _cache;
+    private readonly ILogger? _logger;
+    private readonly IHttpContextAccessor _httpContextAccessor;
+
+    // Constants
+    private const char KEY_DELIMITER = ':';
+
+    /// <summary>
+    /// Create cache service which encapsulates actual caching implementation.
+    /// </summary>
+    /// <param name="cache">Cache implementation.</param>
+    /// <param name="logger">Logger.</param>
+    /// <param name="httpContextAccessor">Accessor which provides httpContext within factory method.</param>
+    public DabCacheService(IFusionCache cache, ILogger<DabCacheService>? logger, IHttpContextAccessor httpContextAccessor)
+    {
+        _cache = cache;
+        _logger = logger;
+        _httpContextAccessor = httpContextAccessor;
+    }
+
+    /// <summary>
+    /// Attempts to fetch response from cache. If there is a cache miss, call the 'factory method' to get a response
+    /// from the backing database.
+    /// </summary>
+    /// <typeparam name="JsonElement">Response payload</typeparam>
+    /// <param name="queryExecutor">Factory method. Only executed after a cache miss.</param>
+    /// <param name="queryMetadata">Metadata used to create a cache key or fetch a response from the database.</param>
+    /// <param name="cacheEntryTtl">Number of seconds the cache entry should be valid before eviction.</param>
+    /// <returns>JSON Response</returns>
+    /// <exception cref="Exception">Throws when the cache-miss factory method execution fails.</exception>
+    public async ValueTask<JsonElement?> GetOrSetAsync<JsonElement>(IQueryExecutor queryExecutor, DatabaseQueryMetadata queryMetadata, int cacheEntryTtl)
+    {
+        string cacheKey = CreateCacheKey(queryMetadata);
+        JsonElement? result = await _cache.GetOrSetAsync(
+               key: cacheKey,
+               async (FusionCacheFactoryExecutionContext<JsonElement> ctx, CancellationToken ct) =>
+               {
+                   // Need to handle undesirable results like db errors or null.
+                   JsonElement? result = await queryExecutor.ExecuteQueryAsync(
+                       sqltext: queryMetadata.QueryText,
+                       parameters: queryMetadata.QueryParameters,
+                       dataReaderHandler: queryExecutor.GetJsonResultAsync<JsonElement>,
+                       httpContext: _httpContextAccessor.HttpContext!,
+                       args: null,
+                       dataSourceName: queryMetadata.DataSource);
+
+                   ctx.Options.SetSize(EstimateCacheEntrySize(cacheKey: cacheKey, cacheValue: result?.ToString()));
```

Comment:
Is the cache entry size variable and can be modified per request?

Reply:
Yes, cache entry size is based on contents of the key (request metadata) and response (db response).

The cache entry size is calculated once the response is received from the db and then inserted into the cache.